### ISSUE TITLE
DX: Dialog on top of dialog is not the best UI

### ISSUE THREAD

**Issue by manolo** [vaadin/vaadin-crud#73]
Some comments during DX related with this
- design with dialogs is strange. I would like to be able to change that appearance.
- Dialog on top of dialog? Don’t like it.

Opening this ticket for discussion @jouni ?

**Comment by jouni** [comment]
One option is to hide the editor dialog while the confirmation dialog is shown. In both cases (discard changes and confirm delete) the expected result is that the editor dialog will remain closed. Only if the user decides they made a mistake, and want to continue editing or cancel the deletion, the editor dialog be “re-opened”.

That is how the confirm dialogs in the Beverage Buddy example app work: https://beverage-buddy-vok.herokuapp.com

**Comment by manolo** [comment]
IIRC at some point it  was in this way in bakery, and we had to change to keep dialog opened.

**Comment by jouni** [comment]
Yeah, I think I remember that. Do you remember the reason why we changed it?

Before we change anything again, I would like to discuss this with the rest of the design team, and perhaps do a quick user test about it.

**Comment by manolo** [comment]
I think the reason was because close events in dialog, when user presses escape or clicks out, were not able to be prevented, hence the dialog was closed before the server could check for dirty, then server opened the confirmation, and in case of cancel, dialog was reopened again. We hacked the dialog-flow implementation so as overlay was able to ask to server before closing.

**Comment by manolo** [comment]
So the issue about nested dialog in BFF, is that there are two behaviours:  open https://bakery-flow.demo.vaadin.com/products, click on a product, modify a field to enable save button, then try to close the dialog by pushing cancel, or by pressing escape, the behavior and UX is different.

Thus, there is a ticket reflecting this and pointing to a couple of discussions in dialog and dialog-flow about the matter. https://github.com/vaadin/bakery-app-starter-flow-spring/issues/671

**Comment by jouni** [comment]
Shouldn’t it be possible to have the web component works so that everything that closes, the dialog sends an event that can be listened for and then Flow (or any custom app code) can call `preventDefault()` and do the server round-trip to check if closing is allowed? Once the round-trip returns the custom code can decide to actually close the dialog or not.

The UX could also be optimized to expect closing to usually go through, and visually close the dialog already after a short delay (something like 300–500ms), but in the exceptional case when closing is prevented it would bring it back.

**Comment by jouni** [comment]
That event seems to exist already: `vaadin-overlay-close`

https://github.com/vaadin/vaadin-overlay/blob/master/src/vaadin-overlay.html#L415-L421

And it also checks `evt.defaultPrevented`, so I believe it should be possible to implement the behavior I described above.

**Comment by manolo** [comment]
> Shouldn’t it be possible to have the web component works so that everything that closes, the dialog sends an event that can be listened for and then Flow

Yes, that is the idea I implemented in dialog-flow: https://github.com/vaadin/vaadin-dialog-flow/commit/59d5fe25d4c51a44c73974d7774353c4c4b3662a example of usage in https://github.com/vaadin/vaadin-dialog-flow/commit/d81dc21dc06e30e0348f07091ff0567dce067add#diff-3c7d49f1d0928ada3eb0942680094193R95

Implementation does not prevent default unless in server side user adds a dialogCloseActionListener meaning that is server responsability to close the dialog.

But I think whether server side or client side closes the editor before opening the confirmation should be a Design decission or a configurable parameter, so as everything is handled by the WC without making the developer add aditional code and preventing unnecesary communication in the wire.
